>## In this You get to know why numpy is faster than the python list? and How?
>## and, Basic Numpy array creations.
> **Hope You Learn Something.**

> Numpy is a python library that basically deals with the numbers, It is very faster than python list and also it takes very less memory to store etc.

## Space Check

In [5]:
from colorama import Fore
import numpy as np

In [6]:
arr = np.arange(1, 10000)
print(Fore.BLACK, "Numpy, Memory for 1 element and 1000 element")
print(Fore.BLUE, f"Value: {arr[2]} Spaced taken for Single element: {arr[0].itemsize} Bytes")
print(Fore.BLUE, "Total space taken for 10000 element: {}".format(arr[0].itemsize * 10000))

 Numpy, Memory for 1 element and 1000 element
 Value: 3 Spaced taken for Single element: 4 Bytes
 Total space taken for 10000 element: 40000


In [7]:
import sys 
li = [i for i in range(1, 10000)]
print(Fore.LIGHTMAGENTA_EX,"Python List, Memory taken for 1 element and 1000 element")
print(f" Value: {li[2]} Space taken Single element: {sys.getsizeof(li[2])}")
print(f" Total space taken for 10000 elements: {sys.getsizeof(li[2]) * 10000}")

 Python List, Memory taken for 1 element and 1000 element
 Value: 3 Space taken Single element: 28
 Total space taken for 10000 elements: 280000


#### Conlusion for space utilization for python list and numpy array
> **For Numpy single element it took only  4 bytes.**

> **For Python List single element it took 27 bytes.**

## Speed Check

In [26]:
### Taking the 10000 elements for Python list and also for Numpy array
import time


# here i am taking 10 lakh to iterated and check the time
size = 1000000

def arr_mul():
    # creating the 2 arrays to multiply
    arr1 = np.arange(size)
    arr2 = np.arange(size)
    # starting time 
    t1 = time.time()
    # just give * this will do multiply element by element ex: arr1[0]*arr2[0], arr1[1] * arr[1] 
    c = arr1 * arr2
    # removing the star time from current time to get the total time
    return time.time()-t1
    
    
def li_mul():
    # Creating the 2 lists to multiply
    a = range(size)
    b = range(size)
    # Starting time to do multiply
    t2 = time.time()
    # iterating throgh list and adding each element individually
    c = [a[i]*b[i] for i in range(size)]
    return time.time()-t2
    
print(f"Numpy Array, Time taken to multiply for 10 lakh elements: {arr_mul()}")
print(f"Python List, Time taken to multiply for 10 lakh elements: {li_mul()}")

Numpy Array, Time taken to multiply for 10 lakh elements: 0.015623331069946289
Python List, Time taken to multiply for 10 lakh elements: 0.3138282299041748


### conclusion of speed 
> Numpy Arrays: You can observe the above time results **numpy taken 0.015 milliseconds,
    To process 10 lakh elements**.

> Python Lists: **This taken 0.3138 milliseconds to process the 10 lakh elements.**

#### So let's start to create the numpy arrays

In [27]:
## np.array(shape, dtype) --> dtype: datatype.
arr1 = np.array(10) # I given 10 it will create an array with single element.
print(f'array with single element: {arr1}, type: {type(arr)}')

## If you want to create array with multiple elements just pass a list.
arr2 = np.array([1,5,7,4,78,4]) # I passed a list it will create array using that list.
print(f'array with mulitple elements: {arr2}, type: {type(arr)}')

## also you can print the array shape using array.shape.
## so i print the shapes of above array's.

print(f'arr1_shape: {arr1.shape} and arr2_shape" {arr2.shape}\n')
print(Fore.RED, """Note: For arr1 the shape is empty, you might be thought shape must be (1, ),
But Once you call np.array() it is an arbitary object (It doesn't have any fixed size) it is 0D array
but however size is 0, np.arra([10])-> 1d, np.array([[10]]) 2D like this you can go upto nD array.\n""")

print(f'arr2 shape is (6) it is a 0D array with 6 columns')

array with single element: 10, type: <class 'numpy.ndarray'>
array with mulitple elements: [ 1  5  7  4 78  4], type: <class 'numpy.ndarray'>
arr1_shape: () and arr2_shape" (6,)

 Note: For arr1 the shape is empty, you might be thought shape must be (1, ),
But Once you call np.array() it is an arbitary object (It doesn't have any fixed size) it is 0D array
but however size is 0, np.arra([10])-> 1d, np.array([[10]]) 2D like this you can go upto nD array.

arr2 shape is (6) it is a 0D array with 6 columns


In [38]:
## Another important is when you are passing the array it need to be same datatype.
## Numpy array only contains the Homogenous elements let's deep dive into dtypes, how can we use.

## I am passing the different datatypes in numpy array let's see how it behave's.
arr3 = np.array([5, 6, 4, 'data_science', 1.4]) # here i passed int, str, float.

## Let's print and observe the output.
print("arr3: ", arr3, """\nYou can observe here when you are giving the int,str,float. It converting all
values into string, whenever we give str with other it will give higher priority.""")

## Let's remove str and give int and float, let's check which it will take as priority.

arr4 = np.array([1, 2, 2.0,3])
print("arr4: ",arr4, "\n As I told the Priorities so we removed string, after it converting into float")

print()

print(Fore.GREEN, "Note: You can mention datatype which you need while creating.")
# lets create array with specified datatype
arr5 = np.array([1, 2, 3, 5], dtype=float) # I given int but converting into float
print('Int to Float', "arr5: ", arr5)

arr6 = np.array([1, 2, 3, 5], dtype=str) # and also as string
print('Int to String', "arr6: ", arr6)

arr7 = np.array(['1', '2', 3, 4.0], dtype=int) # here i give string number and float and converted to int
print('Str and Float to int', "arr7: ", arr7)

print(Fore.RED, """Note: When you are going to convert to int or float, make sure that, Any alphabets 
not be include. You can give number as string ex: '1', '2', not this ex: 'a', 'any'.""")

arr3:  ['5' '6' '4' 'data_science' '1.4'] 
You can observe here when you are giving the int,str,float. It converting all
values into string, whenever we give str with other it will give higher priority.
arr4:  [1. 2. 2. 3.] 
 As I told the Priorities so we removed string, after it converting into float

 Note: You can mention datatype which you need while creating.
Int to Float arr5:  [1. 2. 3. 5.]
Int to String arr6:  ['1' '2' '3' '5']
Str and Float to int arr7:  [1 2 3 4]
 Note: When you are going to convert to int or float, make sure that, Any alphabets 
not be include. You can give number as string ex: '1', '2', not this ex: 'a', 'any'.


In [36]:
## let' create numpy array using some functions, but note: default it created array with float

# It's like python range function, It range we give start,stop,step.

## np.arange(start, end, step)

arr8 = np.arange(1, 100)
print("arr8: ", arr8, "\nCreated array with 1, 100. It will start from 1 it go till 99, start will be inclusive and end will exlusive.")

## we can also create array by giving step arguments.
## ex: start: 1, end: 10, step: 3. It will create with 1, 4,7
arr9 = np.arange(1, 20, 3)
print("arr9: ",arr9, "\nIt created array start with 1 taking step as 3. You can observe array to get better understand.")

arr8:  [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99] 
Created array with 1, 100. It will start from 1 it go till 99, start will be inclusive and end will exlusive.
arr9:  [ 1  4  7 10 13 16 19] 
It created array start with 1 taking step as 3. You can observe array to get better understand.


In [50]:
### np.ones(shape, dtype) -> It return new array with given filled with one.

arr10 = np.ones((2, 4))
print(Fore.GREEN, "Created array with shape (2, 4), 2rows and 4columns. default dtype: float")
print("arr10: ", arr10) 


print(Fore.BLUE, "\nCreated array with shape(5, 10), by give int datatype.")
arr11 = np.ones((5, 10), dtype=int)
print("arr11: ", arr11)
print(Fore.RED, "\nYou can create which shape you want and data type.")

 Created array with shape (2, 4), 2rows and 4columns. default dtype: float
arr10:  [[1. 1. 1. 1.]
 [1. 1. 1. 1.]]
 
Created array with shape(5, 10), by give int datatype.
arr11:  [[1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]]
 
You can create which shape you want and data type.


In [53]:
## np.zeros(shape, dtype) -> This also same like np.ones() but it return array with zeros.

print(Fore.GREEN, "Created array with np.zeros(), input: shape, and dtype: default(float).")
arr12 = np.zeros((2, 14))
print("arr12: ", arr12)

print(Fore.RED, "\n Note: Remaining same like np.ones().")

 Created array with np.zeros(), input: shape, and dtype: default(float).
arr12:  [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
 
 Note: Remaining same like np.ones().


In [73]:
## np.empty(shape, dtype)-> It creates an array by given shape may with 0 or random values.

print(Fore.RED, "\n NOTE: For Better Understanding I created to arrays with same shape and data type.")

print(Fore.GREEN, "\nYou can find one of this array will be zero and one will be random values.")
arr13 = np.empty((5, 6), dtype=int)
print("arr13: ", arr13)

arr14 = np.empty((5, 6), dtype=int)
print("arr14: ", arr14)

 
 NOTE: For Better Understanding I created to arrays with same shape and data type.
 
You can find one of this array will be zero and one will be random values.
arr13:  [[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]
arr14:  [[3801155 5570652 6619251 7536754 7536732 6357101]
 [7602290 3014748 7340137 7602297 7274600 6029422]
 [7471216 6684783 7077993 6226021 6619236 6357094]
 [7078005 6029428 6881384 7602291 7471215 3014777]
 [7405683 6881388 6619252 7798829 7077985       0]]


In [80]:
## np.full(shape, dtype, fill_value), It return new array with given and shape, it filled with given filled value


print(Fore.RED, "Created the array with desired value. Here i passed 76.")
arr15 = np.full((5, 6), dtype=float, fill_value=76)
print(Fore.GREEN, "\narr15: ", arr15)

 Created the array with desired value. Here i passed 76.
 
arr15:  [[76. 76. 76. 76. 76. 76.]
 [76. 76. 76. 76. 76. 76.]
 [76. 76. 76. 76. 76. 76.]
 [76. 76. 76. 76. 76. 76.]
 [76. 76. 76. 76. 76. 76.]]


In [93]:
## np.linspace(start, stop, num=defualt: 50, endPoint=True).
# start: Starting point
# stop: End point
# num: number of samples need to created, if you don't give any number, It will generate 50 default
"""endPoint=True. This is interesting whenever, You create array with np.arange,ones,zeros. In those
functions stop is exlusive, but it np.linspace, it is options by give endPoint=True, The stop point will
be included. If False it will be excluded. """
# Return evenly spaced numbers over a specified interval.
# Returns num evenly spaced samples, calculated over the interval [start, stop].
# The endpoint of the interval can optionally be excluded.

print(Fore.RED, "Created array with end point include, means endPoint=True")
arr16 = np.linspace(1, 5, num=10)
print(Fore.GREEN, "arr16: ", arr16)

print(Fore.RED, "\nCreated array with end point exclude, means endPoint=False and with default num samples 50.")
arr17 = np.linspace(1, 5, endpoint=False)
print(Fore.GREEN, "arr17: ", arr17)


print(Fore.RED, "\nYou can check the difference between points")
print(Fore.GREEN, f"arr16 difference between points: {arr16[2]-arr16[1]}")
print(Fore.GREEN, f"arr17 difference between points: {arr17[5]-arr17[4]}")

 Created array with end point include, means endPoint=True
 arr16:  [1.         1.44444444 1.88888889 2.33333333 2.77777778 3.22222222
 3.66666667 4.11111111 4.55555556 5.        ]
 
Created array with end point exclude, means endPoint=False and with default num samples 50.
 arr17:  [1.   1.08 1.16 1.24 1.32 1.4  1.48 1.56 1.64 1.72 1.8  1.88 1.96 2.04
 2.12 2.2  2.28 2.36 2.44 2.52 2.6  2.68 2.76 2.84 2.92 3.   3.08 3.16
 3.24 3.32 3.4  3.48 3.56 3.64 3.72 3.8  3.88 3.96 4.04 4.12 4.2  4.28
 4.36 4.44 4.52 4.6  4.68 4.76 4.84 4.92]
 
You can check the difference between points
 arr16 difference between points: 0.4444444444444444
 arr17 difference between points: 0.07999999999999985
